### PR TITLE
Web Prompts Styles and Validation

### PR DESCRIPTION
**Pull request by rgomezp**
# Description
## 1 Line Summary
Slidedown Channel Capture container implementation with validation and styling

### Styling
New sass styles for the Web Prompts slidedown. One consideration to keep in mind is the conflict with the [International Telephone Input](https://github.com/jackocnr/intl-tel-input) library styles. This PR includes changes to the "reset styles" that works around this issue. The way it works is by excluding the reset styles from classes that start with "iti."

### Validation
Validation is done for both the sms and email fields. The two work slightly differently. We access the latest validation state for both input types via a `ChannelCaptureContainer` member variable that is updated on every `keyup` or `focus` event. These variables are then accessed on the `ALLOW_CLICK` event to throw an error or continue with the updating of the player records.

#### SMS Validation
Uses the [3rd party library's](https://github.com/jackocnr/intl-tel-input#getting-started-not-using-a-bundler) validation function to update the `smsInputFieldIsValid` variable of the `ChannelCaptureContainer`. 

#### Email Validation
Uses regex to update the `smsInputFieldIsValid` variable of the `ChannelCaptureContainer`. 

### Implementation
New classes:

* `ChannelCaptureContainer`
* `ChannelCaptureError`

#### `ChannelCaptureContainer`
Class that manages the creation and mounting (to DOM) of the HTML elements required to render the Web Prompts slidedown. Uses the WebAPI to create DOM elems.

Other functions:

* Initializes the 3rd party telephone input library
* Mounts event listeners to the native HTML elements that listen for keystrokes
* Logic to validate inputs and style the slidedown accordingly

#### `ChannelCaptureError`
Extends the `OneSignalError` class to provide an easy way to throw errors specific to the `ChannelCaptureContainer`.

# Systems Affected
* [x] WebSDK
* [ ] Backend
* [ ] Dashboard

# Validation
### Checklist
* [ ] All the automated tests pass or I explained why that is not possible
* [ ] I have personally tested this on my machine or explained why that is not possible
* [ ] I have included test coverage for these changes or explained why they are not needed

**Programming Checklist**
Interfaces:

* [x] Don't use default export
* [x] New interfaces are in model files

Functions:

* [x] Don't use default export
* [ ] All function signatures have return types
* [x] Helpers should not access any data but rather be given the data to operate on.

Typescript:

* [x] No Typescript warnings
* [x] Avoid silencing null/undefined warnings with the exclamation point

Other:

* [x] Iteration: refrain from using `elem of array` syntax. Prefer `forEach` or use `map`
* [x] Avoid using global OneSignal accessor for `context` if possible. Instead, we can pass it to function/constructor so that we don't call `OneSignal.context`

### Info
Validation:
Screen Shot 2021-03-16 at 6 00 34 PM

Screen Shot 2021-03-16 at 6 03 39 PM

Screen Shot 2021-03-16 at 6 01 07 PM

### Checklist
* [x] I have included screenshots/recordings of the intended results or explained why they are not needed

- - -
## Related Tickets
- - -
## 
This change is [Reviewable](https://reviewable.io/reviews/onesignal/onesignal-website-sdk/768)

### CODE REVIEW

File: src/slidedown/ChannelCaptureContainer.ts [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+            // show error
+            validationElement.classList.remove(CHANNEL_CAPTURE_CONTAINER_CSS_CLASSES.onesignalValidationElementHidden);
+            inputElement.classList.add(CHANNEL_CAPTURE_CONTAINER_CSS_CLASSES.onesignalErrorInput);
+        } else {
+            validationElement.classList.add(CHANNEL_CAPTURE_CONTAINER_CSS_CLASSES.onesignalErrorInput);
+            inputElement.classList.remove(CHANNEL_CAPTURE_CONTAINER_CSS_CLASSES.onesignalValidationElementHidden);
+        }
+    }
+
+    static resetInputErrorStates() {
```
Comment (rgomezp) [inline src/slidedown/ChannelCaptureContainer.ts:269]: void return type

**Review by jkasten2 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-website-sdk/768)* status: 0 of 10 files reviewed, 2 unresolved discussions (waiting on @rgomezp)

___
*[typings/globals/missing.d.ts, line 32 at r21](https://reviewable.io/reviews/onesignal/onesignal-website-sdk/768#-MW1n-WzE8-ykxv8vg16:-MW1n-X-E1AybZcuQy-8:b-u0esfg) ([raw file](https://github.com/onesignal/onesignal-website-sdk/blob/b138bd1bbdb8d0e558c1e2ad6a8d8af1ca0f0031/typings/globals/missing.d.ts#L32)):*
> ```XML
> 
> declare var OneSignal: any;
> declare var intlTelInput: any;
> ```

I there a index.d.ts defined for this library?

File: package.json [hunk 2]
```diff
@@ -118,7 +118,7 @@
     },
```
Comment (rgomezp) [inline package.json:118]: ALREADY REVIEWED IN https://github.com/OneSignal/OneSignal-Website-SDK/pull/771

File: src/slidedown/InternationalTelephoneInput.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+export enum ItiScriptURLs {
```
Comment (rgomezp) [inline src/slidedown/InternationalTelephoneInput.ts:1]: ALREADY REVIEWED IN https://github.com/OneSignal/OneSignal-Website-SDK/pull/771

File: src/slidedown/ChannelCaptureContainer.ts [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+                `#${CHANNEL_CAPTURE_CONTAINER_CSS_IDS.emailInputWithValidationElement}`
+            );
+            const emailValidationElement = getDomElementOrStub(
+                `#${CHANNEL_CAPTURE_CONTAINER_CSS_IDS.onesignalEmailValidationElement}`
+            );
+            removeCssClass(emailInputWithValidation, CHANNEL_CAPTURE_CONTAINER_CSS_CLASSES.onesignalErrorInput);
+            addCssClass(emailValidationElement,
+                CHANNEL_CAPTURE_CONTAINER_CSS_CLASSES.onesignalValidationElementHidden
+                );
+
```
Comment (rgomezp) [inline src/slidedown/ChannelCaptureContainer.ts:203]: Extra line

File: src/slidedown/ChannelCaptureContainer.ts [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+        document.head.appendChild(script1);
+        document.head.appendChild(script2);
+        document.head.appendChild(link);
+
+        const promise1 = new Promise<void>(resolve => { script1.onload = () => { resolve(); }; });
+        const promise2 = new Promise<void>(resolve => { script2.onload = () => { resolve(); }; });
+
+        await Promise.all([promise1, promise2]);
+    }
```
Comment (rgomezp) [inline src/slidedown/ChannelCaptureContainer.ts:233]: ALREADY REVIEWED IN #769

File: src/slidedown/ChannelCaptureContainer.ts [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+
+    /* P U B L I C */
+    public async mount(): Promise<void> {
+        const isUsingSms = ChannelCaptureContainer.isUsingSmsInputField(this.promptOptions.type);
+        const isUsingEmail = ChannelCaptureContainer.isUsingEmailInputField(this.promptOptions.type);
+
+        if (isUsingSms) {
+            await this.loadPhoneLibraryScripts();
+        }
+
```
Comment (rgomezp) [inline src/slidedown/ChannelCaptureContainer.ts:243]: ALREADY REVIEWED IN #769

File: src/slidedown/ChannelCaptureContainer.ts [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+        body.append(captureContainer);
+
+        if (isUsingSms) {
+            this.initializePhoneInputLibrary();
+            this.addSmsInputEventListeners();
+        }
+
+        if (isUsingEmail) {
+            this.addEmailInputEventListeners();
+        }
```
Comment (rgomezp) [inline src/slidedown/ChannelCaptureContainer.ts:255]: ALREADY REVIEWED IN #769

File: src/slidedown/ChannelCaptureContainer.ts [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+        return re.test(emailString || '') || emailString === "";
+    }
+
+    static isUsingSmsInputField(type: DelayedPromptType): boolean {
+        return type === DelayedPromptType.Sms || type === DelayedPromptType.SmsAndEmail;
+    }
+
+    static isUsingEmailInputField(type: DelayedPromptType): boolean {
+        return type === DelayedPromptType.Email || type === DelayedPromptType.SmsAndEmail;
+    }
```
Comment (rgomezp) [inline src/slidedown/ChannelCaptureContainer.ts:332]: ALREADY REVIEWED IN #769

**Comment by rgomezp** [comment]
___
*[typings/globals/missing.d.ts, line 32 at r21](https://reviewable.io/reviews/onesignal/onesignal-website-sdk/768#-MW1n-WzE8-ykxv8vg16:-MW6A9zDAs7Nu4aEYeZv:bqmjwcm) ([raw file](https://github.com/onesignal/onesignal-website-sdk/blob/b138bd1bbdb8d0e558c1e2ad6a8d8af1ca0f0031/typings/globals/missing.d.ts#L32)):*
<details><summary><i>Previously, jkasten2 (Josh Kasten) wrote…</i></summary><blockquote>

I there a index.d.ts defined for this library?
</blockquote></details>

Added!

File: src/slidedown/ChannelCaptureContainer.ts [hunk 9]
```diff
@@ -137,30 +132,90 @@ export default class ChannelCaptureContainer {
...
+        // TO DO: refactor all this
+        emailInput.addEventListener('keyup', () => {
+            const emailValue = (<HTMLInputElement>emailInput).value;
+            this.emailInputFieldIsValid = ChannelCaptureContainer.validateEmailInputWithReturnVal(emailValue);
+
+            this.updateValidationOnInputChange(DelayedPromptType.Email);
+        });
+    }
+
+    private updateValidationOnInputChange(type: DelayedPromptType): void {
```
Comment (jkasten2) [inline src/slidedown/ChannelCaptureContainer.ts:179]: Should be two methods. `updateValidationOnEmailChange`  and `updateValidationOnSMSChange`. After that we could clean up more by then making a shared function these to can use to update the CSS ect.

**Review by jkasten2 — CHANGES_REQUESTED** [review]

**Pull request merged**